### PR TITLE
Implement `/mv dumps` command

### PR DESCRIPTION
**Pull request by zax71**
## Current changes/additions

 - Renamed `/mv version` to `/mv dumps`
 - Added `--logs` option that uploads logs to https://mclo.gs
 - Added `bukkit.yml` and `server.properties` to dump
 - Added installed plugin list to dump
 - Implemented `--pastebincom` `--pastesdev` `--pastegg` `--logs` and `--exclude-plugin-list`
 - Replace Hastebin with [pastes.dev](https://pastes.dev)

### CODE REVIEW

**Comment by dumptruckman** [comment]
I don't think I'm a fan of completely replacing the version command. I would rather keep that command and maybe change it's functionality - perhaps it logs out some version information or something to the console/sender. Then it could also mention that `/mv dumps` can be used to share information about the plugin/their server.

**Comment by zax71** [comment]
I might add that in another PR, would like to finish this one first though :)

File: src/main/java/com/onarandombox/MultiverseCore/configuration/node/NodeGroup.java [hunk 1]
```diff
@@ -1,15 +1,11 @@
...
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.Map;
-
 import io.github.townyadvanced.commentedconfiguration.setting.CommentedNode;
 import io.vavr.control.Option;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.*;
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/configuration/node/NodeGroup.java:7]: Don't use wildcard imports, intellij has a config for it. I usually just set like to 9999 imports before using wildcards lol

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java [hunk 2]
```diff
@@ -104,23 +83,27 @@ public SimpleMVWorldManager(
...
-        File[] files = server.getWorldContainer().listFiles((file, s) -> s.equalsIgnoreCase("bukkit.yml"));
-        if (files != null && files.length == 1) {
-            FileConfiguration bukkitConfig = YamlConfiguration.loadConfiguration(files[0]);
-            if (bukkitConfig.isConfigurationSection("worlds")) {
-                Set<String> keys = bukkitConfig.getConfigurationSection("worlds").getKeys(false);
-                for (String key : keys) {
-                    defaultGens.put(key, bukkitConfig.getString("worlds." + key + ".generator", ""));
-                }
+
+        if (plugin.getBukkitConfig() == null) {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:94]: Is that change relevant to the /mv dumps command? Also I recall you said getBukkitConfig() is only supported on paper or smt?

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java [hunk 3]
```diff
@@ -1001,24 +985,24 @@ public FileConfiguration getConfigWorlds() {
...
+        }
+        return false;
+    }
 
     /**
      * {@inheritDoc}
      */
     @Override
-	public Collection<String> getPotentialWorlds() {
+    public Collection<String> getPotentialWorlds() {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:1005]: I think all these formatting issue should be a separate pr, I think your IDE prob auto fixed the formatting Bcu you edited the file lol

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                        @NotNull MultiverseCore plugin,
+                        @NotNull MVCoreConfig config,
+                        @NotNull MVWorldManager worldManager) {
+        super(commandManager);
+        this.config = config;
+        this.plugin = plugin;
+        this.worldManager = worldManager;
+
+        registerFlagGroup(CommandFlagGroup.builder("mvdumps")
+                .add(CommandFlag.builder("--pastebincom")
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:53]: I think --pastebin good enough, don't need com

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                .add(CommandFlag.builder("--hastebin")
+                        .addAlias("-h")
+                        .build())
+                .add(CommandFlag.builder("--pastegg")
+                        .addAlias("-p")
+                        .build())
+                .add(CommandFlag.builder("--logs")
+                        .addAlias("-l")
+                        .build())
+                .add(CommandFlag.builder("--include-plugin-list")
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:68]: I actually had the thought, plugin list should be included by default. And have a flag to exclude it.

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+        BukkitRunnable logPoster = new BukkitRunnable() {
+            @Override
+            public void run() {
+                HashMap<String, String> pasteURLs = new HashMap<>();
+                boolean hasArgs = false;
+
+                if (parsedFlags.hasFlag("--pastebincom")) {
+                    hasArgs = true;
+                    sendMessage(issuer, "Uploading debug info to https://pastebin.com");
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:110]: Prob good idea to make all the sendMessage string into i18n instead

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java (hunk 3 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:1005]: Yep, that was the case lol

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 4 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:53]: I did think that originally but I don’t people using —pagebin to put it on a pastebon, not realising it actually means pastebin.com

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 5 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:68]: Nice idea. I’ll implement that one

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 6 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:110]: Adding to the todo

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java (hunk 2 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:94]: I needed the Bukkit.yml file and server.properties for the dump so I felt it should be abstracted there too. It also makes it much easier to edit when the issue is fixed

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 4 above)
Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:65]: huh then what it means?

Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:63]: Upload to https://pastebin.com

File: src/main/java/com/onarandombox/MultiverseCore/MultiverseCore.java [hunk 7]
```diff
@@ -330,6 +330,34 @@ public String getAuthors() {
         return authors.toString();
     }
 
+    @Nullable
+    public File getBukkitConfig() {
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/MultiverseCore.java:334]: I'm not crazy about adding these methods to MultiverseCore after I worked to move most of our API stuff out of it. Is this functionality that is going to be useful elsewhere? Instead, we could inject the Server into the command class and just make these private methods in there.

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+                    String link = pasteURLs.get(service);
+                    issuer.sendInfo(MVCorei18n.DUMPS_URL_LIST, "{service}", service, "{link}", link);
+                }
+
+            }
+        };
+
+        // Run the uploader async as it could take some time to upload the debug info
+        logPoster.runTaskAsynchronously(plugin);
+    }
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:201]: This method is nearly 90 lines long (way too long). Could we extract some of this, like especially the runnable stuff, into a separate method/class?

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java [hunk 9]
```diff
@@ -103,23 +103,27 @@ public SimpleMVWorldManager(
...
+            Logging.warning("Any Default worldgenerators will not be loaded!");
+            return;
+        }
+
+        FileConfiguration bukkitConfig = YamlConfiguration.loadConfiguration(bukkitConfigFile);
+
+        if (bukkitConfig.isConfigurationSection("worlds")) {
+            Set<String> keys = bukkitConfig.getConfigurationSection("worlds").getKeys(false);
+            for (String key : keys) {
+                defaultGens.put(key, bukkitConfig.getString("worlds." + key + ".generator", ""));
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:125]: We might want to wait until this is dealt with in the new world manager...

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+        // Deal with default case of upload
+        Services services = parsedFlags.flagValue("--upload", Services.class);
+        if (services == null) {
+            services = Services.PASTEGG;
+            Logging.finer("Upload was null so set to pastegg");
+        }
+
+
+        // Need to be final for some reason...
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:152]: This is bcu you reference the variable in anonymous function

Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:152]: Ah, ty for clearing that up!

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+                        pasteURLs.put("paste.gg", postToService(PasteServiceType.PASTEGG, true, null, files));
+                    }
+                    case PASTESDEV -> {
+                        issuer.sendInfo(MVCorei18n.DUMPS_UPLOADING, "{link}", "https://pastes.dev");
+                        pasteURLs.put("pastes.dev", postToService(PasteServiceType.PASTESDEV, true, null, files));
+                    }
+                }
+
+                // Finally, loop through and print all URLs
+                for (String service : pasteURLs.keySet()) {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:191]: I dont really see a need for loop, bcu its just printing 1 or 2 things

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        final Services finalServices = services;
+        BukkitRunnable logPoster = new BukkitRunnable() {
+            @Override
+            public void run() {
+                HashMap<String, String> pasteURLs = new HashMap<>();
+                Logging.finer("Logs type is: " + finalLogsType);
+                Logging.finer("Services is: " + finalServices);
+
+                // Deal with logs flag
+                if (!parsedFlags.hasFlag("--paranoid")) {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:163]: I think this check should be done above with the `// Deal with default case of logs`

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+                        .build())
+                .build());
+    }
+
+    private enum Services {
+        PASTEGG,
+        PASTESDEV
+    }
+
+    private enum LogsType {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:105]: `LogsTypeOption` maybe is clearer?

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+                            return types;
+                        })
+                        .build())
+                .add(CommandFlag.builder("--paranoid")// Does not upload logs or plugin list (except if --logs mclogs is there)
+                        .addAlias("-p")
+                        .build())
+                .build());
+    }
+
+    private enum Services {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:100]: `SevicesTypeOption` maybe is clearer?

**Review by benwoo1110 — CHANGES_REQUESTED** [review]

File: src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java (hunk 9 above)
Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/world/SimpleMVWorldManager.java:127]: I think its fine,  new world manager wont be touching generators stuff anytime soon

File: src/main/java/com/onarandombox/MultiverseCore/utils/MVCorei18n.java [hunk 15]
```diff
@@ -37,6 +37,11 @@ public enum MVCorei18n implements MessageKeyProvider {
     DELETE_SUCCESS,
     DELETE_PROMPT,
 
+    // Dumps command
+    DUMPS_DESCRIPTION,
+    DUMPS_UPLOADING,
+    DUMPS_UPLOADING_LOGS,
+    DUMPS_URL_LIST,
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/utils/MVCorei18n.java:44]: new line after this is preferred

File: src/main/java/com/onarandombox/MultiverseCore/configuration/node/NodeGroup.java [hunk 16]
```diff
@@ -10,6 +10,7 @@
 import java.util.Iterator;
 import java.util.Map;
 
+
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/configuration/node/NodeGroup.java:13]: why a random new line?

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+
+        // Deal with default case of logs
+        LogsType logsType = parsedFlags.flagValue("--logs", LogsType.class);
+        if (logsType == null) {
+            logsType = LogsType.MCLOGS;
+            Logging.finer("Logs was null so set to mclogs");
+        }
+
+
+
```
[the remark this answers is not in the document]
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:144]: Sorry... I always forget to fix it lol

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 11 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:188]: Good point. I might just make it print out directly

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+
+        // Deal with default case of upload
+        Services services = parsedFlags.flagValue("--upload", Services.class);
+        if (services == null) {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:144]: set a defaultValue for command flag, so you dont need null checking

**Review by benwoo1110 — COMMENTED** [review]
I think overall the logic checking can be more readable instead of scatter at various parts of the code, for example don't reuse hasFlag for the same flag multiple times:

```java
boolean paranoid = parsedFlags.hasFlag("--paranoid");
Services services = parsedFlags.flagValue("--upload", Services.class);
LogsType logsType = parsedFlags.flagValue("--logs", LogsType.class);
if (!paranoid && logsType == null) {
     logsType = LogsType.MCLOGS;
}
```

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 8 above)
Comment (zax71) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:191]: Regarding moving `logPoster` to its own class, where would that be placed as it is only really used in `DumpsCommand.java` so `utils/` wouldn't make sense imo.

File: src/main/java/com/onarandombox/MultiverseCore/MultiverseCore.java [hunk 19]
```diff
@@ -330,6 +330,8 @@ public String getAuthors() {
         return authors.toString();
     }
 
+
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/MultiverseCore.java:333]: ahhh whitespace

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+
+        BukkitRunnable logPoster = new BukkitRunnable() {
+            @Override
+            public void run() {
+                HashMap<String, String> pasteURLs = new HashMap<>();
+                Logging.finer("Logs type is: " + logsType);
+                Logging.finer("Services is: " + servicesType);
+
+                // Deal with logs flag
+                if (!parsedFlags.hasFlag("--paranoid")) {
```
Comment (benwoo1110) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:150]:
```suggestion
                if (!paranoid) {
```

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 8 above)
Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java:181]: I think separating to methods is fine. Looks to me like you can split "Deal with logs flag" and "Deal with uploading debug info" into separate methods.

**Review by benwoo1110 — CHANGES_REQUESTED** [review]

File: src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java (hunk 18 above)
Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DumpsCommand.java]: ohhh right you can do `parsedFlags.flagValue("--upload", ServiceOption.PASTEGG, ServiceOption.class)`

**Review by benwoo1110 — APPROVED** [review]
